ase/io/espresso_cp.py

```diff
@@ -76,9 +76,44 @@
 def write_espresso_cp_in(fd, atoms, input_data=None, pseudopotentials=None,
                       kspacing=None, kpts=None, koffset=(0, 0, 0),
                       **kwargs):
+
     write_espresso_in(fd, atoms, input_data, pseudopotentials,
                       kspacing, kpts, koffset, **kwargs)
 
+    if not fd.closed:
+        fd.close()
+    
+    # Extra blocks
+    extra_lines = []
+    input_parameters = construct_namelist(input_data, **kwargs)
+    for section in input_parameters:
+        if section.lower() not in ['ee', 'nksic']:
+            continue
+        extra_lines.append('&{0}\n'.format(section.upper()))
+        for key, value in input_parameters[section].items():
+            if value is True:
+                extra_lines.append('   {0:16} = .true.\n'.format(key))
+            elif value is False:
+                extra_lines.append('   {0:16} = .false.\n'.format(key))
+            elif value is not None:
+                # repr format to get quotes around strings
+                extra_lines.append('   {0:16} = {1!r:}\n'.format(key, value))
+        extra_lines.append('/\n')  # terminate section
+    extra_lines.append('\n')
+
+    # Read in the original file without the NKSIC and EE blocks
+    with open(fd.name, 'r') as fd_read:
+        lines = fd_read.readlines()
+
+    # Find where to insert the extra blocks
+    i_break = lines.index('\n')
+    before = lines[:i_break]
+    after = lines[i_break:]
+
+    # Rewrite the file with the extra blocks
+    with open(fd.name, 'w') as fd_rewrite:
+        fd_rewrite.writelines(before + extra_lines + after)
```

Comment:
We need to insert these blocks mid-way through the file, so unfortunately what we have to do is read in the file and overwrite it. It's a bit messy but it's the simplest solution I could think of that means that the `CP` modifications don't leak into the `PW` code

Reply:
The upshot of this is that actually there is an alternative method for `kpts = "exclude"` -- we could simply remove the `kpts` block from `after` before overwriting the file

Reply:
I'm not sure if this is a preferable approach to `kpts = "exclude". What do you think @degennar ?

Reply:
The advantage of this way as I see it is that we don't have to alter `espresso.py` for the sake of `CP`

Reply:
Ok, I guess you also removed the `EE` and `NKSIC` namelists from `espresso.py` then.
Yes, since I guess the best is to keep `espresso.py` as untouched as possible I would say that your solution is better.